File: apps/api/internal/assistant/confirmation_allowlist.go
package assistant

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/Ciao1019/Petrichor/apps/api/ent"
	"github.com/Ciao1019/Petrichor/apps/api/ent/assistantthread"
)

const dangerAllowlistTTL = 24 * time.Hour

var destructiveCriticalTools = map[string]struct{}{
	"delete_article":        {},
	"delete_document":       {},
	"delete_ai_config":      {},
	"revoke_article_share":  {},
	"revoke_agent_api_key":  {},
	"set_public_qa_enabled": {},
}

type dangerAllowlistState struct {
	ToolNames []string `json:"toolNames"`
	UpdatedAt string   `json:"updatedAt"`
}

func IsDestructiveCriticalTool(toolName string) bool {
	_, ok := destructiveCriticalTools[toolName]
	return ok
}

func parseDangerAllowlistJSON(raw *string) *dangerAllowlistState {
	if raw == nil || strings.TrimSpace(*raw) == "" {
		return nil
	}
	var state dangerAllowlistState
	if err := json.Unmarshal([]byte(*raw), &state); err != nil {
		return nil
	}
	if len(state.ToolNames) == 0 || state.UpdatedAt == "" {
		return nil
	}
	return &state
}

func isAllowlistExpired(state *dangerAllowlistState, now time.Time) bool {
	t, err := time.Parse(time.RFC3339, state.UpdatedAt)
	if err != nil {
		return true
	}
	return now.Sub(t) > dangerAllowlistTTL
}

func getThreadDangerAllowlist(ctx context.Context, client *ent.Client, threadID, userID int64) (map[string]struct{}, error) {
	if threadID <= 0 {
		return map[string]struct{}{}, nil
	}
	row, err := client.AssistantThread.Query().
		Where(
			assistantthread.IDEQ(threadID),
			assistantthread.UserIDEQ(userID),
			assistantthread.DeletedAtIsNil(),
		).
		Only(ctx)
	if err != nil {
		if ent.IsNotFound(err) {
			return map[string]struct{}{}, nil
		}
		return nil, err
	}
	state := parseDangerAllowlistJSON(row.DangerAllowlistJSON)
	if state == nil {
		return map[string]struct{}{}, nil
	}
	if isAllowlistExpired(state, time.Now().UTC()) {
		_ = clearThreadDangerAllowlist(ctx, client, threadID, userID)
		return map[string]struct{}{}, nil
	}
	out := make(map[string]struct{}, len(state.ToolNames))
	for _, name := range state.ToolNames {
		if name != "" {
			out[name] = struct{}{}
		}
	}
	return out, nil
}

func isToolInThreadDangerAllowlist(ctx context.Context, client *ent.Client, threadID, userID int64, toolName string) (bool, error) {
	if IsDestructiveCriticalTool(toolName) {
		return false, nil
	}
	set, err := getThreadDangerAllowlist(ctx, client, threadID, userID)
	if err != nil {
		return false, err
	}
	_, ok := set[toolName]
	return ok, nil
}

func addToolToThreadDangerAllowlist(ctx context.Context, client *ent.Client, threadID, userID int64, toolName string) error {
	if threadID <= 0 || IsDestructiveCriticalTool(toolName) {
		return nil
	}
	set, err := getThreadDangerAllowlist(ctx, client, threadID, userID)
	if err != nil {
		return err
	}
	set[toolName] = struct{}{}
	names := make([]string, 0, len(set))
	for name := range set {
		names = append(names, name)
	}
	for i := 0; i < len(names); i++ {
		for j := i + 1; j < len(names); j++ {
			if names[j] < names[i] {
				names[i], names[j] = names[j], names[i]
			}
		}
	}
	next := dangerAllowlistState{
		ToolNames: names,
		UpdatedAt: time.Now().UTC().Format(time.RFC3339),
	}
	raw, _ := json.Marshal(next)
	_, err = client.AssistantThread.Update().
		Where(assistantthread.IDEQ(threadID), assistantthread.UserIDEQ(userID), assistantthread.DeletedAtIsNil()).
		SetDangerAllowlistJSON(string(raw)).
		Save(ctx)
	return err
}

func clearThreadDangerAllowlist(ctx context.Context, client *ent.Client, threadID, userID int64) error {
	_, err := client.AssistantThread.Update().
		Where(assistantthread.IDEQ(threadID), assistantthread.UserIDEQ(userID)).
		ClearDangerAllowlistJSON().
		Save(ctx)
	return err
}

func dangerousToolWhitelist(toolName string) (string, bool) {
	m := map[string]string{
		"delete_article":               "article.delete",
		"revoke_article_share":         "share.revoke",
		"delete_document":              "document.delete",
		"delete_ai_config":             "ai_config.delete",
		"update_ai_config_credentials": "ai_config.update_credentials",
		"revoke_agent_api_key":         "agent_api_key.revoke",
		"set_public_qa_enabled":        "public_qa.set_enabled",
	}
	v, ok := m[toolName]
	return v, ok
}

func isDangerousToolName(toolName string) bool {
	_, ok := dangerousToolWhitelist(toolName)
	return ok
}

func assertConfirmationAction(toolName string) error {
	if _, ok := dangerousToolWhitelist(toolName); !ok {
		return fmt.Errorf("不允许确认执行未知危险工具：%s", toolName)
	}
	reg := GetToolRegistration(toolName)
	if reg == nil || reg.Risk != "dangerous" {
		return fmt.Errorf("危险工具未注册：%s", toolName)
	}
	return nil
}

File: apps/web/src/server/assistant/confirmation-allowlist.ts
import { and, eq, isNull } from "drizzle-orm"
import { getDb } from "@/server/db/client"
import { assistantThreads } from "@/server/db/schema"

export const DANGER_ALLOWLIST_TTL_MS = 24 * 60 * 60 * 1000

/**
 * 永不进入会话 allowlist 的破坏性工具（每次仍需确认卡）。
 * 对齐计划：delete_* / revoke_* / set_public_qa_enabled。
 */
export const DESTRUCTIVE_CRITICAL_TOOL_NAMES = new Set([
    "delete_article",
    "delete_document",
    "delete_ai_provider",
    "revoke_article_share",
    "revoke_agent_api_key",
    "set_public_qa_enabled",
])

export type DangerAllowlistState = {
    toolNames: string[]
    updatedAt: string
}

export function isDestructiveCriticalTool(toolName: string): boolean {
    return DESTRUCTIVE_CRITICAL_TOOL_NAMES.has(toolName)
}

export function parseDangerAllowlistJson(raw: string | null | undefined): DangerAllowlistState | null {
    if (!raw?.trim()) return null
    try {
        const parsed = JSON.parse(raw) as unknown
        if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) return null
        const record = parsed as { toolNames?: unknown; updatedAt?: unknown }
        if (!Array.isArray(record.toolNames) || typeof record.updatedAt !== "string") return null
        const toolNames = record.toolNames.filter((item): item is string => typeof item === "string" && item.length > 0)
        return { toolNames, updatedAt: record.updatedAt }
    } catch {
        return null
    }
}

export function isAllowlistExpired(state: DangerAllowlistState, now = Date.now()): boolean {
    const updatedAt = Date.parse(state.updatedAt)
    if (!Number.isFinite(updatedAt)) return true
    return now - updatedAt > DANGER_ALLOWLIST_TTL_MS
}

/** 读取线程 allowlist；过期则清空并返回空。 */
export async function getThreadDangerAllowlist(input: {
    threadId: number
    userId: number
}): Promise<Set<string>> {
    const [thread] = await getDb()
        .select({ dangerAllowlistJson: assistantThreads.dangerAllowlistJson })
        .from(assistantThreads)
        .where(and(
            eq(assistantThreads.id, input.threadId),
            eq(assistantThreads.userId, input.userId),
            isNull(assistantThreads.deletedAt),
        ))
        .limit(1)

    const state = parseDangerAllowlistJson(thread?.dangerAllowlistJson)
    if (!state) return new Set()
    if (isAllowlistExpired(state)) {
        await clearThreadDangerAllowlist(input)
        return new Set()
    }
    return new Set(state.toolNames)
}

export async function isToolInThreadDangerAllowlist(input: {
    threadId: number
    userId: number
    toolName: string
}): Promise<boolean> {
    if (isDestructiveCriticalTool(input.toolName)) return false
    const names = await getThreadDangerAllowlist(input)
    return names.has(input.toolName)
}

export async function addToolToThreadDangerAllowlist(input: {
    threadId: number
    userId: number
    toolName: string
}): Promise<void> {
    if (isDestructiveCriticalTool(input.toolName)) return
    const existing = await getThreadDangerAllowlist(input)
    existing.add(input.toolName)
    const next: DangerAllowlistState = {
        toolNames: [...existing].sort(),
        updatedAt: new Date().toISOString(),
    }
    await getDb()
        .update(assistantThreads)
        .set({
            dangerAllowlistJson: JSON.stringify(next),
            updatedAt: new Date(),
        })
        .where(and(
            eq(assistantThreads.id, input.threadId),
            eq(assistantThreads.userId, input.userId),
            isNull(assistantThreads.deletedAt),
        ))
}

export async function clearThreadDangerAllowlist(input: {
    threadId: number
    userId: number
}): Promise<void> {
    await getDb()
        .update(assistantThreads)
        .set({
            dangerAllowlistJson: null,
            updatedAt: new Date(),
        })
        .where(and(
            eq(assistantThreads.id, input.threadId),
            eq(assistantThreads.userId, input.userId),
        ))
}
